nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java

```diff
@@ -0,0 +1,126 @@
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_384.toString()),
+                            MessageDigestAlgorithms.SHA_384)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_512.toString()),
+                            MessageDigestAlgorithms.SHA_512)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.RSASSA_PSS_SHA_256.toString()),
+                            MessageDigestAlgorithms.SHA_256)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.RSASSA_PSS_SHA_384.toString()),
+                            MessageDigestAlgorithms.SHA_384)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.RSASSA_PSS_SHA_512.toString()),
+                            MessageDigestAlgorithms.SHA_512)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.ECDSA_SHA_256.toString()),
+                            MessageDigestAlgorithms.SHA_256)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.ECDSA_SHA_384.toString()),
+                            MessageDigestAlgorithms.SHA_384)
+                    .put(JWSAlgorithm.parse(SigningAlgorithmSpec.ECDSA_SHA_512.toString()),
+                            MessageDigestAlgorithms.SHA_512)
+                    .build();
+
+    /**
+     * The supported JWS algorithms (alg).
+     * <p>
+     * Note: We are using KMS prescribed algorithm names here.
+     * Ref: https://docs.aws.amazon.com/kms/latest/developerguide/symm-asymm-choose.html#key-spec-rsa-sign
+     */
+    public static final Set<JWSAlgorithm> SUPPORTED_ALGORITHMS = JWS_ALGORITHM_TO_MESSAGE_DIGEST_ALGORITHM.keySet();
+
+    protected KmsAsymmetricSigningCryptoProvider(
+            @NonNull final AWSKMS kms, @NonNull final String privateKeyId, @NonNull final MessageType messageType) {
+        super(SUPPORTED_ALGORITHMS);
+        this.kms = kms;
+        this.privateKeyId = privateKeyId;
+        this.messageType = messageType;
+    }
+
+    protected ByteBuffer getMessage(final JWSHeader header, final byte[] signingInput) throws JOSEException {
+        final var alg = header.getAlgorithm();
+        var message = signingInput;
+
+        String messageDigestAlgorithm = Optional.ofNullable(JWS_ALGORITHM_TO_MESSAGE_DIGEST_ALGORITHM.get(alg))
+                .orElseThrow(() -> new JOSEException(
+                        String.format("No digest algorithm exist for the JWS algorithm %s in map: %s",
+                                alg, JWS_ALGORITHM_TO_MESSAGE_DIGEST_ALGORITHM)));
+
+        if (messageType == MessageType.DIGEST) {
+            MessageDigest messageDigestProvider;
+            try {
+                messageDigestProvider = MessageDigest.getInstance(messageDigestAlgorithm);
+            } catch (NoSuchAlgorithmException e) {
+                throw new JOSEException("Invalid message digest algorithm.", e);
+            }
+            message = messageDigestProvider.digest(message);
+        }
```

Comment:
[Might be taken as follow up re-work]
This behaviour doesn't seem ideal to me from a user's perspective. 
Internally I don't believe there is any difference in the way hashing is done, either by `messageDigestProvider` or hashing done by KMS. Why would clients care at which layer the hashing is being done. They can only pass RAW unhashed inputs here. 

Wouldn't it make more sense to rather allow them to pass both kinds of messages (`RAW`/`DIGEST`) and then just pass that to KMS?

Reply:
One counter argument can be that we'll rely on clients to insure that the passed messageType is correct. Can't we have validations for such a case?